core-bundle/src/DependencyInjection/Compiler/RegisterFragmentsPass.php

```diff
@@ -83,6 +83,14 @@ protected function registerFragments(ContainerBuilder $container, string $tag):
         $command = $container->hasDefinition('contao.command.debug_fragments') ? $container->findDefinition('contao.command.debug_fragments') : null;
 
         foreach ($this->findAndSortTaggedServices($tag, $container) as $reference) {
+            // If a controller has multiple methods for different fragment type (e.g. a content element
+            // and a front end module), the first RegisterFragmentsPass creates a child definition that
+            // inherits all tags from the original. On the next run, the RegisterFragmentsPass would
+            // pick up the child definition and try to create duplicate fragments.
+            if (0 === strpos((string) $reference, 'contao.fragment._')) {
```

Comment:
Wouldn't it be safer to keep track of an array of fragments for which the child definition has already been created?

Reply:
that was actually my first draft, but it is much more complicated because
 1. that's not a service, we create multiple instances in the bundle
 2. we would need to _reset_ that, but again, not being a service makes that complicated.

Reply:
This prefix thing feels shaky to me. Why is there a second run anyway? Every pass is only passed once?

Reply:
A controller has two tags:
 - `contao.content_element`
 - `contao.frontend_module`

The bundle runs the pass for each fragment type
https://github.com/contao/contao/blob/master/core-bundle/src/ContaoCoreBundle.php#L89-L103

The original tags are duplicated to the child definitions because we need other tags (e.g. for monolog logger)
https://github.com/contao/contao/blob/master/core-bundle/src/DependencyInjection/Compiler/RegisterFragmentsPass.php#L117

Reply:
Ah, now I think I've understood the problem. Hmm, so the `RegisterFragmentsPass` itself is not a service which means we cannot store the fact that we've already created a child definition. But the container is. What do you think about adding another tag and checking for this then? E.g. when done creating the child class, calling `addTag(self::DUPLICATE_CHECK_TAG_NAME)` and check for this instead of a service id prefix?

Reply:
I don't see the point. The service ID prefix is unique to services generated by the pass, it would be a problem if its not. Adding a tag does not make it "more unique", anyone could abuse the tag the same way they can abuse the prefix …

Reply:
It's just nicer. Anyway, I'm fine with the prefix as well. However, we need it to be a class constant and tests.